xapian-applications/omega/query.cc

```diff
@@ -651,24 +652,27 @@ html_strip(const string &str)
     return res;
 }
 
-// FIXME split list into hash or map and use that rather than linear lookup?
-static int word_in_list(const string& word, const string& list)
-{
-    string::size_type split = 0, split2;
-    int count = 0;
-    while ((split2 = list.find('\t', split)) != string::npos) {
-	if (word.size() == split2 - split) {
-	    if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-		return count;
-	}
-	split = split2 + 1;
-	++count;
-    }
-    if (word.size() == list.size() - split) {
-	if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-	    return count;
+static unordered_map<string, int> word_to_occurrence;
+
+static void build_word_map(const string& list) {
+    string word;
+    int word_index = 0;
+    for (size_t i = 0; i <= list.length(); ++i) {
+	if (list[i] == '\t' || list[i] == '\0') {
+	    if (word_to_occurrence.count(word) == 0) {
+		word_to_occurrence[word] = word_index++;
+		word.clear();
+	    }
+	} else
+	    word += list[i];
     }
-    return -1;
+}
+
+static int word_in_list(const string& word) {
+    int pos = -1;
+    if (word_to_occurrence.count(word))
+	pos = word_to_occurrence[word];
```

Comment:
> That comment is unhelpfully hidden by github because it's on a bit of code which has since been modified again - you need to click on "Hide outdated" under the second of James' comments to see that.

Thanks, I see it now. Addressing that comment, I'll replace `count()` with `find()` everywhere in my changes.

> The code you quote doesn't search once in each case - it searches once in the case where word isn't present, but twice in the case where it is 

Sorry, I keep forgetting the extra linear time overhead due to `operator []`. I think it's because I've used it several times under the impression of its average case time complexity which is constant. Working on this PR has changed my perception towards it completely, especially when working on performance-centric applications in real world